### PR TITLE
[API][SIMS-#681]MSFAA send file

### PR DESCRIPTION
**Pull request by guru-aot**
Tasks:
- [x] Create a new reference application id column in the MSFAA number table and insert it while a new MSFAA number is created
- [x] Fetch all the MSFAA numbers, where the requested date is null. Fetch the referenced application, student, offering, and institution details to create the records in the MSFAA send file records.
- [x] Create sequence control for full time, part time files separately and reset the sequence every day to generate the files.
- [x] Create sequence control for file.
- [x] Using similar process to CRA, create fixed fixed file and send to zoneB SFTP server
- [x] Populate column date_requested in MSFAANumbers table

image

image

### CODE REVIEW

File: sources/packages/api/src/services/application/application.service.e2e-spec.ts
Comment (guru-aot) [inline sources/packages/api/src/services/application/application.service.e2e-spec.ts:184]: Removed it, as the related msfaa number has been removed during the testCompletedApplication removed in line 183

File: sources/packages/api/src/msfaa-integration/msfaa-validation.service.ts
Comment (guru-aot) [inline sources/packages/api/src/msfaa-integration/msfaa-validation.service.ts:142]: For now, hard coded to married, as this will be taken care when we create a new column in the application table as part of another story.

File: sources/packages/api/src/database/entities/msfaa-number.model.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/database/entities/msfaa-number.model.ts:88]: Should it be marked as optional with "_referenceApplication?_"?

File: sources/packages/api/src/database/entities/msfaa-number.model.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/database/entities/msfaa-number.model.ts:83]: _Cascade_ must be set to false, right?

File: sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts:54]: Please remove the extra line.

File: sources/packages/api/src/route-controllers/msfaa-integration/msfaa-integration.system.controller.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/msfaa-integration/msfaa-integration.system.controller.ts:28]: Do you see any issue with executing both in parallel?

File: sources/packages/api/src/database/entities/msfaa-number.model.ts
Comment (guru-aot) [reply sources/packages/api/src/database/entities/msfaa-number.model.ts:88]: We agreed to have it as mandatory field, as this should be available when we create a application that generated a New MSFAA Number.

File: sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts
Comment (ann-aot) [inline sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts:123]: a general qn, if the server is Linux, will it ignore `carriage return` or consider it as `line feed`(i.e ` \n\n`)?

File: sources/packages/api/src/msfaa-integration/msfaa-request.service.ts
Comment (ann-aot) [inline sources/packages/api/src/msfaa-integration/msfaa-request.service.ts:133]: why `maritalStatus` is static

File: sources/packages/api/src/utilities/msfaa-utils.ts
Comment (ann-aot) [inline sources/packages/api/src/utilities/msfaa-utils.ts:9]: this can also be simplified if you would like to, return maritalStatus === "married" ? "M" : maritalStatus === "single" ? "S" : "O";

**Review by ann-aot — COMMENTED** [review]
LGTM 👍 Just some minor comments

File: sources/packages/api/src/msfaa-integration/msfaa-request.service.ts
Comment (guru-aot) [reply sources/packages/api/src/msfaa-integration/msfaa-request.service.ts:133]: https://github.com/bcgov/SIMS/pull/827#discussion_r740505848 - For now we are having it static, this will be changed when we decide on where we are fetching the marital status in the future stories. I believe I have mentioned it in my comments.

File: sources/packages/api/src/database/entities/msfaa-number.model.ts
Comment (andrewsignori-aot) [reply sources/packages/api/src/database/entities/msfaa-number.model.ts:88]: Agree, it must be mandatory, but on the migrations it was optional. So, as we talked, we are going to be changing the migrations in an upcoming ticket to the column be mandatory on DB also.

File: sources/packages/api/src/utilities/msfaa-utils.ts
Comment (guru-aot) [reply sources/packages/api/src/utilities/msfaa-utils.ts:9]: I tried it and code smells picked it and said "Ternary operators should not be nested". So changed it.
https://rules.sonarsource.com/javascript/RSPEC-3358

File: sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts
Comment (guru-aot) [reply sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts:123]: I am not sure, but I think it will not be ignored and in our case, a line is fed

File: sources/packages/api/src/route-controllers/msfaa-integration/msfaa-integration.system.controller.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/msfaa-integration/msfaa-integration.system.controller.ts]: This comment seems to be incomplete and it is also mentioning "to validate if the MSFAA number", which is not accurate.

File: sources/packages/api/src/route-controllers/msfaa-integration/msfaa-integration.system.controller.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/msfaa-integration/msfaa-integration.system.controller.ts:19]: I believe that "request the MSFAA number" is not accurate. We are the ones generating the number and sending it, right?

File: sources/packages/api/src/route-controllers/msfaa-integration/msfaa-integration.system.controller.ts
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/msfaa-integration/msfaa-integration.system.controller.ts]: Thanks for doing the change 😉

File: sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts
Comment (andrewsignori-aot) [reply sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts:123]: For the CRA migration, the break lines were defined as _"\r\n"_ and the files were properly received by the CRA. We follow the same that was being done on PROD files.
@abschwenker can confirm on PROD files the expectation for MSFAA also.

Comment (abschwenker) [reply sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts:123]: correct, these files use the same mechanisms to be sent. So the way the CRA files are sent can be mirrored

File: sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts
Comment (dheepak-aot) [inline sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts:122]: commented line of code.

File: sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts
Comment (dheepak-aot) [inline sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts:164]: Just to know, why are we using dynamic repo here?

File: sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts
Comment (guru-aot) [reply sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts:122]: https://github.com/bcgov/SIMS/pull/827#discussion_r740505848 - For now we are having it commented, this will be changed when we decide on where we are fetching the marital status in the future stories. I believe I have mentioned it in my comments.

File: sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts
Comment (guru-aot) [reply sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts:164]: Added in the comments, 
   * @param [externalRepo] when provided, it is used instead of the
   * local repository (this.repo). Useful when the command must be executed,
   * for instance, as part of an existing transaction manage externally to this
   * service.

File: sources/packages/api/src/utilities/msfaa-utils.ts
Comment (dheepak-aot) [reply sources/packages/api/src/utilities/msfaa-utils.ts:9]: IMO if there is more that 2 marital status possible, switch case would be nice.

File: sources/packages/api/src/utilities/msfaa-utils.ts
Comment (dheepak-aot) [inline sources/packages/api/src/utilities/msfaa-utils.ts:19]: same as previous comment about switch case.

File: devops/openshift/api-deploy.yml
Comment (dheepak-aot) [inline devops/openshift/api-deploy.yml:175]: Just to know why is name of key : ${MSFAA_REQUEST_FOLDER_NAME_KEY} is API_SECRET_NAME

Correct me if I am missing something

File: sources/packages/api/src/msfaa-integration/msfaa-request.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/msfaa-integration/msfaa-request.service.ts:34]: We have the enum OfferingIntensity for this that is actually the one that is being used while calling this method.
Can you please change it to be the enum type?

File: sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts
Comment (dheepak-aot) [reply sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts:164]: Sorry @guru-aot I couldn't understand from the comments still. That is why asked.

File: sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts
Comment (dheepak-aot) [reply sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts:122]: Okay. seeing this it made me feel that we commented out a property which we didn't need to be returned at last minute. Got it now.

File: sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts:155]: It is supposed to be named as msfaaRequestIds,

File: sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts
Comment (andrewsignori-aot) [reply sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts:122]: Whatever it is it would be good to add a TODO comment. This will be marked on Sonar Cloud and we will never forget it 😉

File: sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts:110]: Please use the enum.

**Comment by dheepak-aot** [comment]
Nice work @guru-aot 👍. Minor comments.

File: sources/packages/api/src/msfaa-integration/msfaa-request.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/msfaa-integration/msfaa-request.service.ts:135]: IMO, accessing the same index over and over again does not seem right. I would say that it is clear when we access the index once, assign it to a variable and then access the properties from this variable.

File: sources/packages/api/src/msfaa-integration/msfaa-request.service.ts
Comment (andrewsignori-aot) [reply sources/packages/api/src/msfaa-integration/msfaa-request.service.ts:133]: Please add a TODO comment to mark it and make SonarCloud highlight it.

File: sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts:84]: It could be simplified as below
```
msfaaDetail.addressLine2 = msfaaRecord.addressLine2 ??  "";
```

**Review by andrewsignori-aot — COMMENTED** [review]
What a great work, congrats 😉 
Please take a look at the comments.

File: sources/packages/api/src/msfaa-integration/msfaa-request.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/msfaa-integration/msfaa-request.service.ts:80]: As we talked about, this method is creating a new transaction that will not fail together with the rest of this block.
As we discussed, lets try to pass the entityManager to the nested consumeNextSequence to make everything part of the same transation.

File: sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts
Comment (ann-aot) [reply sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts:122]: https://superuser.com/questions/374028/how-are-n-and-r-handled-differently-on-linux-and-windows, If it's tested in both servers, then, I am good

File: sources/packages/api/src/services/sequence-control/sequence-control.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/sequence-control/sequence-control.service.ts:87]: Can we move this parameter to the second position? I believe that it would be cleaner when it is being invoked. Does it make sense?
The _process_ parameter could be big enough to make the _entityManager_ be lost a bit at the end of the list of parameters when calling it.

File: sources/packages/api/src/services/sequence-control/sequence-control.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/sequence-control/sequence-control.service.ts:130]: This method is no longer executing a rollback, right?
I would say that the log could be removed, which will leave only "throw error;" that does not make sense then.
Maybe the entire try/catch could be removed or please just change the error log.

File: sources/packages/api/src/utilities/msfaa-utils.ts
Comment (andrewsignori-aot) [reply sources/packages/api/src/utilities/msfaa-utils.ts:9]: > IMO if there is more that 2 marital status possible, switch case would be nice.

😄 it was exactly what I also mentioned in a call.

**Review by ann-aot — APPROVED** [review]
LGTM 👍 Good work @guru-aot. regarding CL and LF, if it was tested and passed in both Linux and windows, then I am Good

**Review by abschwenker — APPROVED** [review]
lgtm, once merged I will run a few tests

File: sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts
Comment (andrewsignori-aot) [reply sources/packages/api/src/msfaa-integration/msfaa-integration.service.ts:122]: @ann-aot, they don't need to be tested on both servers, they will not be used by two different servers.
The files have one way to be generated and we are mirroring the PROD files. These files are going to be passed to Windows servers but the final destination are the federal servers that, we have no idea what they are, we do not need to know, we just need to know how the PROD files looks like and follow it 😉 
Hope that it is clear.

File: sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts
Comment (andrewsignori-aot) [reply sources/packages/api/src/services/msfaa-number/msfaa-number.service.ts]: You missed the parameter name in the comments. Please update the case.

File: sources/packages/api/src/msfaa-integration/msfaa-request.service.ts
Comment (andrewsignori-aot) [reply sources/packages/api/src/msfaa-integration/msfaa-request.service.ts:81]: Good work refactoring it and extending the functionality 😉

**Review by dheepak-aot — APPROVED** [review]

**Review by andrewsignori-aot — APPROVED** [review]
Amazing work, thanks for doing the changes 😉

**Pull request merged**
